Add flag to verify OIDC issuer in certificate

haydentherapper opened the pull request:
<!--
Thanks for opening a pull request!

Please remember to:
- mention any issue(s) that this PR closes using a closing keyword as well as the issue number, such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ", cf.
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!
Thank you :)
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

This PR adds support for --cert-oidc-issuer to verify the OIDC issuer of a Fulcio certificate. In combination with --cert-email, users should be able to verify and pin the expected identity of the Fulcio certificate.

Also added certificate generation test utilities. These might be more useful in sigstore/sigstore.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/sigstore/YYYYYY/issues/XXXXX

-->

Ref #556

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->
```release-note
* Added `--cert-oidc-issuer` to verify the OIDC issuer of a Fulcio certificate
```

dlorenc commented:
Nice!

@mattmoor FYI for the policy task force

mattmoor commented:
Awesome.  Looks like it needs a rebase

mattmoor commented on cmd/cosign/cli/options/certificate.go:
```diff
@@ -0,0 +1,40 @@
+	cmd.Flags().StringVar(&o.Cert, "cert", "",
+		"path to the public certificate")
+
+	cmd.Flags().StringVar(&o.CertEmail, "cert-email", "",
+		"the email expected in a valid Fulcio certificate")
```
We should consider changing this since only a handful of things use email, and the rest use URIs.

Doesn't have to be here (probably shouldn't be), but something related that's been bothering me.

mattmoor replied:
A lot of that info is in extensions already (I think you added it!), and I think this is where it may make more sense to pass some payload with all the data to something like Cue/Rego for policy may be more manageable.

haydentherapper replied:
Left a comment in the bug, I think a cert subject flag would be great to handle both email and URIs.

Given that there is ongoing work for policies (correct?), I agree that we should probably avoid adding complex policy checks to flags, and keep flag-based checks simple.

mattmoor commented on cmd/cosign/cli/options/certificate.go:
```diff
@@ -0,0 +1,40 @@
+		"the email expected in a valid Fulcio certificate")
+
+	cmd.Flags().StringVar(&o.CertOidcIssuer, "cert-oidc-issuer", "",
+		"the OIDC issuer expected in a valid Fulcio certificate, e.g. https://accounts.google.com or https://oauth2.sigstore.dev/auth")
```
I'd probably bias towards Github as an example here because I believe it's disproportionately the biggest issuer in our logs now.

haydentherapper replied:
Good to know! Changed the first to `https://token.actions.githubusercontent.com`

mattmoor reviewed (commented):
This is great to see 👍

haydentherapper commented:
Rebased and updated with flag doc change. I also removed `x509.ExtKeyUsage(x509.KeyUsageDigitalSignature)` from the list of EKUs used during cert verification, because digital signature is not an extended key usage.
